### ISSUE TITLE
[issue] panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by JeffreyShran** [projectdiscovery/nuclei#242]
**Describe the bug**
I was running a test to see if I could run most of the templates with a single call to a url. This is the custom workflow I ran by echoing in a single url to nuclei.

```
id: unguided
info:
  name: Workflow to run most of the templates
  author: Jeffrey Shran

variables:
  cves: cves/
  default_credentials: default-credentials/
  dns: dns/
  files: files/
  generic_detections: generic-detections/
  panels: panels/
  security_misconfiguration: security-misconfiguration/
  subdomain_takeover: subdomain-takeover/
  technologies: technologies/
  tokens: tokens/
  vulnerabilities: vulnerabilities/

logic:
  |
  cves()
  default_credentials()
  dns()
  files()
  generic_detections()
  panels()
  security_misconfiguration()
  subdomain_takeover()
  technologies()
  tokens()
  vulnerabilities()
```

The command I ran is as follows:

```echo "https://example.com" | nuclei -c 200 -t ~/unguided.yaml -o example_com.nuclei.unguided```

Nuclei runs for 30-45 seconds then produces the error in the screenshot below.

**Nuclei version**
Current Version: 2.1.0

**Screenshot of the error or bug**
![image](https://user-images.githubusercontent.com/10995361/90645125-7e625280-e22d-11ea-936a-9cd0837f703e.png)

**Comment by bauthard** [comment]
```
goroutine 9 [running]:
github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).writeOutputHTTP(0xc0005ca2a0, 0xc000242750, 0xc0004cdb90, 0xc00093e800, 0xe0, 0xc00019fe60, 0x0, 0x0, 0x0)
	/Users/bauthard/Documents/GitHub/nuclei/v2/pkg/executer/output_http.go:68 +0xf7
github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).handleHTTP(0xc0005ca2a0, 0x1849320, 0xc0001a5b00, 0xc000121480, 0x3a, 0xc000242750, 0xc000a0b5f8, 0xc000a0b770, 0x0, 0x6fc23ac00)
	/Users/bauthard/Documents/GitHub/nuclei/v2/pkg/executer/executer_http.go:217 +0x41e
github.com/projectdiscovery/nuclei/v2/pkg/executer.(*HTTPExecuter).ExecuteHTTP(0xc0005ca2a0, 0x1849320, 0xc0001a5b00, 0xc000121480, 0x3a, 0x0, 0xc0004a1950, 0xc0004a1980, 0x0, 0x0, ...)
	/Users/bauthard/Documents/GitHub/nuclei/v2/pkg/executer/executer_http.go:137 +0x304
github.com/projectdiscovery/nuclei/v2/pkg/workflows.(*NucleiVar).Call(0xc0000700a0, 0x0, 0x0, 0x0, 0xc000700028, 0x0, 0x7ff, 0x0)
	/Users/bauthard/Documents/GitHub/nuclei/v2/pkg/workflows/var.go:73 +0x6b2
github.com/d5/tengo/v2.(*VM).run(0xc000700000)
	/Users/bauthard/Documents/go/pkg/mod/github.com/d5/tengo/v2@v2.6.0/vm.go:633 +0x3326
github.com/d5/tengo/v2.(*VM).Run(0xc000700000, 0x1787dd0, 0xc000125140)
	/Users/bauthard/Documents/go/pkg/mod/github.com/d5/tengo/v2@v2.6.0/vm.go:77 +0xcf
github.com/d5/tengo/v2.(*Compiled).RunContext.func1(0xc000355980, 0xc000700000)
	/Users/bauthard/Documents/go/pkg/mod/github.com/d5/tengo/v2@v2.6.0/script.go:222 +0x2b
created by github.com/d5/tengo/v2.(*Compiled).RunContext
	/Users/bauthard/Documents/go/pkg/mod/github.com/d5/tengo/v2@v2.6.0/script.go:221 +0x22d
```

**Comment by vzamanillo** [comment]
It is caused by `rack-mini-profiler.yaml` but I don't know why, yet.

**Comment by vzamanillo** [comment]
> It is caused by `rack-mini-profiler.yaml` but I don't know why, yet.

Not really it is caused because `colorizer` is `nil`, then `writeOutputHTTP` fails when there is a template match.

**Comment by JeffreyShran** [comment]
Nothing to add, but I just wanted to say thanks for helping to debug the issue @vzamanillo - I wouldn't know where to begin!

**Comment by bauthard** [comment]
This should be fixed in master code @JeffreyShran

**Comment by JeffreyShran** [comment]
Massive thanks! Will try it out asap.

**Comment by JeffreyShran** [comment]
@bauthard Is it sufficient to run `GO111MODULE=on go get -u -v github.com/projectdiscovery/nuclei/v2/cmd/nuclei` and then retest? I'm getting the same error?

Should I update in a different way? I installed via the same command on Debian.

**Comment by bauthard** [comment]
@JeffreyShran To pull master code, you can use followings:- 
```
GO111MODULE=on go get -u -v github.com/projectdiscovery/nuclei/v2/cmd/nuclei@master
```
OR use Github install method:- 

```
https://github.com/projectdiscovery/nuclei#from-github
```

**Comment by JeffreyShran** [comment]
@bauthard Yes! Thank you (again!) Seems to be running fine now.

**Comment by JeffreyShran** [comment]
Ended up running out of memory after a while. There's about 8k lines of debug/error info starting like the below.

For info, I'm using a 4GB RAM VPS. Realistically I am asking it to do too much. Shouldn't there be some sort of queuing or memory management in place within golang that handles this though?

I'm not asking for a fix here unless its straightforward. I feel like its my fault for trying to do too much.

```
fatal error: runtime: out of memory                                                                                                                              [37095/45073]

runtime stack:
runtime.throw(0xb6975c, 0x16)
        /usr/local/go/src/runtime/panic.go:1116 +0x72
runtime.sysMap(0xc0d0000000, 0x4000000, 0x111fa58)
        /usr/local/go/src/runtime/mem_linux.go:169 +0xc5
runtime.(*mheap).sysAlloc(0x110afc0, 0x800000, 0x110afc8, 0x400)
        /usr/local/go/src/runtime/malloc.go:715 +0x1cd
runtime.(*mheap).grow(0x110afc0, 0x400, 0x0)
        /usr/local/go/src/runtime/mheap.go:1286 +0x11c
runtime.(*mheap).allocSpan(0x110afc0, 0x400, 0x100, 0x111fa68, 0x7f4596db8da0)
        /usr/local/go/src/runtime/mheap.go:1124 +0x6a0
runtime.(*mheap).alloc.func1()
        /usr/local/go/src/runtime/mheap.go:871 +0x64                                                                                                                          runtime.(*mheap).alloc(0x110afc0, 0x400, 0xc029190101, 0xc000000d80)
        /usr/local/go/src/runtime/mheap.go:865 +0x81                                                                                                                          runtime.largeAlloc(0x7ffe00, 0x7f457d4d0101, 0x460301)
        /usr/local/go/src/runtime/malloc.go:1152 +0x92                                                                                                                        runtime.mallocgc.func1()
        /usr/local/go/src/runtime/malloc.go:1047 +0x46                                                                                                                        runtime.systemstack(0xc000000d80)
        /usr/local/go/src/runtime/asm_amd64.s:370 +0x66
runtime.mstart()                                                                                                                                                                      /usr/local/go/src/runtime/proc.go:1041

goroutine 2200085 [running]:
runtime.systemstack_switch()
        /usr/local/go/src/runtime/asm_amd64.s:330 fp=0xc031165118 sp=0xc031165110 pc=0x462180
runtime.mallocgc(0x7ffe00, 0xa47460, 0x1, 0x0)
        /usr/local/go/src/runtime/malloc.go:1046 +0x895 fp=0xc0311651b8 sp=0xc031165118 pc=0x40e415
```

**Comment by bauthard** [comment]
Can you add the details about the number of **hosts** and **templates** you were running with this?  @JeffreyShran

**Comment by vzamanillo** [comment]
How much free memory do you have before running `nuclei`? are there any other concurrent processes running at the same time?

**Comment by JeffreyShran** [comment]
@bauthard I was pushing it to see what could be done. yaml file below. I was running it on 2738 hosts.

@vzamanillo At idle -  RAM 180MB/3.8GB  CPU 3.1%. No I ran it without anything else running in the background. (other than the usual system stuff)

```
id: do-the-lot
info:
  name: Workflow to run all templates
  author: @JeffreyShran

variables:
  cves: cves
  default_credentials: default-credentials
  dns: dns
  files: files
  generic_detections: generic-detections
  panels: panels
  payloads: payloads
  security_misconfiguration: security-misconfiguration
  subdomain_takeover: subdomain-takeover
  technologies: technologies
  tokens: tokens
  vulnerabilities: vulnerabilities
  workflows: workflows

# Comment out below what you don't want to call.
logic:
  |
  cves()
  default_credentials()
  dns()
  files()
  generic_detections()
  panels()
  payloads()
  security_misconfiguration()
  subdomain_takeover()
  technologies()
  tokens()
  vulnerabilities()
  workflows()
```

**Comment by vzamanillo** [comment]
I have been doing tests on my machine with multiple urls and I have not managed to get past a memory consumption of 23MB, it is also true that this depends a lot on each url and the size of the responses, but it does not look like a memory leak.

**Comment by JeffreyShran** [comment]
mine jumps to 1GB usage within a few seconds and climbing up to over 2GB now. I was using -pbar but have turned that off in favour of -silent and am retesting now. The silent flag seems to be being ignored and Nuclei is printing findings to the screen.

Command `nuclei -c 200 -t ~/nuclei-workflows/unguided.yaml -o example_com.nuclei.unguided -silent`

I can supply my url list if you provide an email?

**Comment by vzamanillo** [comment]
> mine jumps to 1GB usage within a few seconds and climbing up to over 2GB now. I was using -pbar but have turned that off in favour of -silent and am retesting now. The silent flag seems to be being ignored and Nuclei is printing findings to the screen.
> 
> Command `nuclei -c 200 -t ~/nuclei-workflows/unguided.yaml -o example_com.nuclei.unguided -silent`
> 
> I can supply my url list if you provide an email?

Oh, interesting, I will contact you by DM on discord.

**Comment by vzamanillo** [comment]
Confirmed on the first 5s. after running v2.1, thanks @JeffreyShran

`$ cat urls.txt | ./nuclei -c 200 -t ./unguided.yml -o example_com.nuclei.unguided`

![leak](https://user-images.githubusercontent.com/10209695/91284411-262ad380-e78c-11ea-9400-0b61d925b744.png)

**Comment by vzamanillo** [comment]
The problem is located at `ProcessWorkflow` it creates the template list per URL on stdin.

Anyway, even if the templates are cached, the memory consumption would be high (about 600mb in this case).

**Comment by vzamanillo** [comment]
Fixed on #247
